### ISSUE TITLE
let's make clientview info more useful

### ISSUE THREAD

**Issue by phritz** [rocicorp/replicache#53]
when developing bindings one wants to know what happened during sync. however during pull we
- return no information about the request from the client to the diffserver and
- only return clientviewinfo if the request from diffserver to data layer was successful. and not, for example, if that connection was refused.

beginsync should return more complete information about the pull, including:
- status code and error if any in request to diffserver. we should return this information even if the connection was refused or we couldnt send the request for some reason.
- clientviewinfo should include an error message if it doesn't send the request or if it fails for some other reason, eg no client view url configured

**Comment by aboodman** [comment]
This is more or less required for #57.

**Comment by aboodman** [comment]
I suppose. The way it is working now is somewhat by design -- it's following the philosophy that the bindings should do as little as possible, and afaik, the bindings don't actually need this information. afaik, all they *need* to know is:

1. Did sync succeed (where succeed == all the servers responded successfully and the sync protocol was followed)
2. Is reauth needed
3. Was there some other internal, unexpected error

In any case, more important to me is rationalizing the error handling strategy in sync across push and pull. Right now, both push and pull mix errors from class (1) and (3), but in different ways: push puts both errors into the returned struct, and pull puts them both into the returned error.

Also the signature for push is very different from standard Go practice and also from all the other rpc's.

**Comment by aboodman** [comment]
PS separating class (1) and (3) errors blocks fixing #57 properly.

**Comment by aboodman** [comment]
OK looking at this from the starting point of what `BeginSync` and `MaybeEndSync` (collectively, hereafter *sync*) should return.

IMO there are the following classes of results that the bindings *need* from sync:

1. invalid args: user called the API wrong, or at a wrong time
  - We will print out a nice error telling the user what they did wrong to the console
2. internal error: there's a bug in Replicache or some lower layer. Something very unexpected broke
  - We will print out "internal error" to the user console (perhaps with additional inscrutable info). User will complain on twitter and we will ask for the logs.
3. The resulting synchead/root (might not have changed from previous value)
4. Whether the cache is "connected" or not:
  - The cache is "connected" if all of following are true:
    - We got a valid response from diff-server
    - diff-server got a valid response from client view
    - If we tried to contact batch endpoint, we got a valid response with no temporary errors
  - Otherwise the cache is "disconnected"
  - We will refect the connectness in the bindings, and expose an event for when it changes
  - Users will reflect this in their UI
5. Additional human-readable information about why the cache is disconnected, if it is
  - We will print this out to the console
6. Whether reauth is needed (if a call to client view or batch failed with Unauthorized)
7. Any failure or informational messages from the batch endpoint

(1) and (2) are exclusive, and also common to all our RPCs. The others are non-exclusive. They could actually all happen together. (1) and (2) also should be reflected as genuine errors in bindings (e.g., an exception in JavaScript/Flutter). The others should not be errors. One because we don't want them to seem that loud, and two because philosophically, they are expected parts of running this code.

Plan: (1) and (2) will be expressed using the standard Go comma-error idiom. This is nice because Gomobile does the right thing and translates this is into the right kind of platform-specific error, and also generally because tools understand this convention.

(4), (5), and (6) will be expressed by returning detailed structured information about the results of various substeps of sync (see #53). Each bindings implementation will have to reimplement the logic to turn this info into reauth, onlineness, etc., but it seems small and we need the structured data anyway to print to the console.

We will also have to make Push() and Pull() follow this philosophy.

**Comment by phritz** [comment]
> The cache is "connected" if all of following are true:
> ...
> If we tried to contact batch endpoint, we got a valid response with no temporary errors

Since we are not requiring that the batch endpoint be synchronous and the response is purely informational this should probably just be "if we tried to contact the batch endpoint and got a 200 with a valid response of any kind".

**Comment by aboodman** [comment]
Hm, help me understand...

The batch endpoint can *choose* to return synchronous temporary errors, and if it does choose to, it seems like it typically means that some server-side resources is unavailable and these mutations aren't getting applied. I feel like that counts as disconnected.

**Comment by phritz** [comment]
Yeah I guess it is a definitional question. What you are saying sounds to me more like whether it is making progress, which is a stronger statement to me than being connected. The problem in my mind is that you can't actually tell if it is making progress by just looking at the push response. An async data layer we just talked about, but also consider some kind of problem where the batch endpoint accepts mutations but the client view is stuck at some mutation id and never sees them. It will apparently be making progress according to the above, but won't actually be. 

I don't feel strongly about this, and can get behind "it is connected if mutations are being accepted by the batch endpoint", I just wanted point out the space for misinterpretation.

I guess if we wanted to determine whether it is actually making progress (and not just that we can round trip) we could stuff nop commits in periodically and ensure they come back in a timely fashion.

**Comment by aboodman** [comment]
Agree. It's kind of weird too because client view and batch handler can fail independently. According to above logic we might say cache is disconnected, but user still sees new updates (from other users)!

This is one downside of separating the two endpoints ... like it's technically more robust, but from a user pov maybe more confusing.
